### PR TITLE
wasm: add basic docs

### PR DESCRIPTION
**Pull request by kyessenov**
Change-Id: I35f8f86bfa165d415e4fbc9f3a7c037e968cac25
Commit Message: add wasm overview
Risk Level: low
Testing: none
Docs Changes: yes

### CODE REVIEW

**Comment by phlax** [comment]
@kyessenov headsup that you dont need to regenerate the `/docs` link - it redirects to the most recent build automatically

File: docs/root/intro/arch_overview/advanced/wasm.rst [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+<envoy_v3_api_field_extensions.wasm.v3.PluginConfig.root_id>` field value), and how to route the configuration call to
+the extension during the xDS update.
+
+
+Execution model for filters
+---------------------------
+
+At runtime, the Wasm modules are executed inline in the worker threads via a series of stream callbacks as defined by
+the ABI. The worker threads operate independently and do not share the Wasm execution instances and their runtime
+memory.  Asychronous operations such as sub-requests are accomplished via the host delegating calls and the subsequent
```
Comment (botengyao) [inline docs/root/intro/arch_overview/advanced/wasm.rst:36]: s/Asychronous/Asynchronous

File: docs/root/intro/arch_overview/advanced/wasm.rst [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+(`C++ <https://github.com/proxy-wasm/proxy-wasm-cpp-sdk>`_, `Rust <https://github.com/proxy-wasm/proxy-wasm-rust-sdk>`_) and
+run anywhere. Please see the :ref:`sandbox example <install_sandboxes_wasm_filter>` for instructions on how to create and deploy
+a Wasm module.
+
+Contexts, plugins, and VMs
+--------------------------
+
+A Wasm context refers to the interface implementation between the host (Envoy) and the guest (Wasm bytecode). For an
+HTTP filter, there are two types of contexts: a *root context* (context with ID 0) represents the configuration-only
+context, and a *per-request context* that is created once per each request.
```
Comment (botengyao) [inline docs/root/intro/arch_overview/advanced/wasm.rst:18]: is this context per plug-in? Maybe also mention that since a wasm VM can contain multiple plugins, so it can have multiple root contexts?

**Review by botengyao — COMMENTED** [review]
Thanks @kyessenov for adding this! LGTM module a typo.

File: docs/root/intro/arch_overview/advanced/wasm.rst [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+Wasm
+====
+
+Envoy supports execution of the Wasm modules implemented against the `Proxy-Wasm specification <https://github.com/proxy-wasm/spec>`_.
+Currently, the `binary interface version 0.2.1 <https://github.com/proxy-wasm/spec/tree/main/abi-versions/v0.2.1>`_ is recommended.
+Wasm offers a portable and compact binary executable format that can be compiled once from a variety of languages
+(`C++ <https://github.com/proxy-wasm/proxy-wasm-cpp-sdk>`_, `Rust <https://github.com/proxy-wasm/proxy-wasm-rust-sdk>`_) and
+run anywhere. Please see the :ref:`sandbox example <install_sandboxes_wasm_filter>` for instructions on how to create and deploy
+a Wasm module.
+
```
Comment (mpwarres) [inline docs/root/intro/arch_overview/advanced/wasm.rst:12]: Suggest adding a short section on the different types of plugins, with links to their config protos:

```
Envoy provides several extension points at which Wasm plugins can be invoked:

* As an :ref:`HTTP filter <envoy.extensions.filters.http.wasm.v3.Wasm>`
* As a :ref:`network-level (L4) filter <envoy.extensions.filters.network.wasm.v3.Wasm>`
* As a :ref:`StatsSink <envoy.extensions.stat_sinks.wasm.v3.Wasm>`
* As an :ref:`AccessLogger <envoy.extensions.access_loggers.wasm.v3.WasmAccessLog>`
* As a :ref:`background service <envoy.extensions.wasm.v3.WasmService>`

The particular functions that Envoy invokes on a Wasm plugin depend on the
extension point at which it is configured.
```

File: docs/root/intro/arch_overview/advanced/wasm.rst [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+Contexts, plugins, and VMs
```
Comment (mpwarres) [inline docs/root/intro/arch_overview/advanced/wasm.rst:13]: Suggest mentioning the ability to choose between Wasm runtimes:

```
Wasm runtimes

Envoy Wasm can be :ref:`configured <envoy.extensions.wasm.v3.VmConfig.runtime>`
to use one of several Wasm runtime implementations: V8, WAMR, or
Wasmtime. There is also a special pseudo-Wasm runtime, called the "Null VM", in
which Wasm plugin code is compiled to native (non-Wasm) code and linked
directly into the Envoy binary.

```

File: docs/root/intro/arch_overview/advanced/wasm.rst [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+--------------------------
+
+A Wasm context refers to the interface implementation between the host (Envoy) and the guest (Wasm bytecode). For an
+HTTP filter, there are two types of contexts: a *root context* (context with ID 0) represents the configuration-only
+context, and a *per-request context* that is created once per each request.
+
+A Wasm VM refers to the execution instance of a Wasm bytecode. Multiple VMs can be created for the same bytecode when
```
Comment (mpwarres) [inline docs/root/intro/arch_overview/advanced/wasm.rst:20]: Saying that a Wasm VM is the execution instance of Wasm bytecode could be confusing when applied to multiple Wasm plugins. How about rewording to:

```
A Wasm VM refers to an instance of a Wasm runtime that executes Wasm
plugins. When configuring Wasm plugins, the vm_id field can be used to control
whether plugins run in the same Wasm VM as each other, or separate ones. VM
sharing optimizes the runtime overhead and reduces the total binary size of the
bytecode since the extensions can share the code.

```

File: docs/root/intro/arch_overview/advanced/wasm.rst (hunk 1 above)
Comment (mpwarres) [inline docs/root/intro/arch_overview/advanced/wasm.rst:36]: Typo: Asynchronous

"sub-requests" might not be clear, how about "Asynchronous or blocking operations initiated by plugins, such as
outbound HTTP/gRPC calls, are delegated to Envoy to perform; upon completion, Envoy then invokes a separate completion callback on the plugin."

**Review by mpwarres — COMMENTED** [review]
Thanks for creating this! LGTM mod a few minor comments.

**Comment by mpwarres** [comment]
Tagging @wbpcode in case he also has any comments

File: docs/root/intro/arch_overview/advanced/wasm.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (mpwarres) [reply docs/root/intro/arch_overview/advanced/wasm.rst:24]: Ack, SGTM. (I'd mark this resolved but for some reason the github UI does not give me that option)

File: docs/root/intro/arch_overview/advanced/wasm.rst (hunk 4 above)
Comment (mpwarres) [reply docs/root/intro/arch_overview/advanced/wasm.rst:25]: SG, I agree that positioning this later is better.

**Review by mpwarres — APPROVED** [review]
Thanks for writing this!

**Review by wbpcode — APPROVED** [review]
